### PR TITLE
Use Imatcher

### PR DESCRIPTION
**Pull request by palkan**
Use [Imatcher](https://github.com/teachbase/imatcher.git) for comparison.

This commit only replaces current comparison implementation with Imatcher's one, very straightforward.
A lot of work is to be done: configuration, parameters passing, more tests, etc.

I've also removed old Ruby versions from travis.yml, 'cause they are not supported by Imatcher.

P.S. And now tests run twice faster)

### CODE REVIEW

**Comment by ronen** [comment]
@palkan have you looked into the build failure?

**Comment by palkan** [comment]
I've got no idea, why it fails. Imatcher itself works well with JRuby.

**Comment by jekuno** [comment]
@ronen I finished testing right now. Is it fine for you if I merge this PR?

File: lib/rspec/page-regression/matcher.rb [hunk 1]
```diff
@@ -9,10 +9,15 @@ module RSpec::PageRegression
...
-      opt = args.select { |k,_| RENDER_ARGS.include?(k) }
-
-      Renderer.render_responsive(page, @responsive_filepaths, opt)
-      @comparisons = @responsive_filepaths.map{ |filepaths| ImageComparison.new(filepaths) }
+      render_options = args.select { |k,_| RENDER_ARGS.include?(k) }
+      compare_options = {}
+      compare_options[:threshold] = args.fetch(:threshold, RSpec::PageRegression.threshold)
+      compare_options[:mode] = args.fetch(:mode, RSpec::PageRegression.mode)
+      compare_options[:include_rect] = args.fetch(:include_area, nil)
+      compare_options[:exclude_rect] = args.fetch(:exclude_rect, nil)
```
Comment (ronen) [inline lib/rspec/page-regression/matcher.rb]: The README above says "you can use all of matcher's options", but these lines hardwire support for exactly these 4 options.  I'd suggest adding another argument `imatcher: {options: 'would', go: 'here'}`, but with these four options supported as shortcuts.  that'd be easy to implement, just replace

```    compare_options={}```

with

```    compare_options = args.fetch(:imatcher, {})```

(And of course document that in the README.)  What do you think?

**Comment by ronen** [comment]
@jekuno all looks good to me, but see my comment above re an `imatcher:` arg

File: lib/rspec/page-regression/matcher.rb (hunk 1 above)
Comment (palkan) [reply lib/rspec/page-regression/matcher.rb]: I like the idea, will do

Comment (jekuno) [reply lib/rspec/page-regression/matcher.rb]: @ronen I see your point. However for consistency I'd prefer to either have all imatcher args wrapped in an imatcher hash or have all imatcher args flattened into the main args hash. As we have a well defined dependency to imatcher (`gem 'imatcher', '~> 0.1.9'`) it would be fine for me to hardcode all known imatcher options of the according imatcher version. This also protects us from users who try to use imatcher options which are incompatible with our gem (nobody can foresee which options might be supported by imatcher in future).

Also we should replace `include_area` by `include_rect` in line 16.

Comment (palkan) [reply lib/rspec/page-regression/matcher.rb]: Done.

Comment (jekuno) [reply lib/rspec/page-regression/matcher.rb]: @palkan @ronen Currently there's a check to only allow use of `ALLOWED_ARGS` (see https://github.com/teachbase/rspec-page-regression/blob/2f6f9fc4147c9d788ac3bf34d45c19a7050d3a24/lib/rspec/page-regression/matcher.rb#L54). That means users can in any case only use imatcher options which have been whitelisted explicitly. I'd suggest to keep a comprehensive list of all supported imatcher args in `IMATCHER_ARGS` and to merge the values of the used imatcher args into the main hash of `compare_options` (instead of nesting some args into an `imatcher` hash and placing other args (such as `threshold`) at top level).

For details also see https://github.com/rprt/rspec-page-regression/pull/31/files/a4e386fcd0924c91c29cc67ca346e5cf6a90250b#r100282420 
What do you think?

**Review by jekuno — CHANGES_REQUESTED** [review]

File: lib/rspec/page-regression/matcher.rb (hunk 1 above)
Comment (palkan) [reply lib/rspec/page-regression/matcher.rb]: Totally agree.

Comment (ronen) [reply lib/rspec/page-regression/matcher.rb]: @jekuno here's my reasoning for going with the way that I initially suggested:

- _Reasons for hardwired args for the common cases:_  

  * So that typical users of this gem don't also need to learn imatcher in order to use it
  * To  protect those users from possible future changes to imatcher args.

- _Reason to support_ `:imatcher` _hash arg:_ 
   *  To allow users who do want to take advantage of other features of imatcher to be able to do so.  
   * Regarding fear of imatcher args that are incompatible, I wouldn't worry about that:; any user sophisticated enough to use fancier imatcher features would be sophisticated enough.
   *  `:imatcher` would be in `ALLOWED_ARGS` so no further whitelisting would be needed.

- _Reasons **not** to hardcode all imatcher options_:  

     * It would require duplication of documentation (or send users to read imatcher doc)
     * It would be extra work to maintain, with specs needed for each arg.
     * It would add more work to update this gem whenever imatcher updates.
     * It introduces unneeded coupling between versions.  We actually *shouldn't* depend on `~> 0.1.9` we should depend on `~> 0.1`.  In that way that anyone using both imatcher and this gem would be able to get bug fixes and new features of imatcher without waiting for us to update this gem (note we have a bad track record of waiting a year to fix issues...!). 

       [@palkan I assume that you're not following the pointless (IMHO) semver rule that allows backwards-incompatible changes in minor releases if major release is 0.  If so, please update to 1.0!]

      * It leaves an opening for potential naming conflicts (in the future) between args native to this gem and imatcher arg.  E.g. supposing we decide to come up with an arg `:strict` that has some impact on how spec names are mapped to filenames, and later imatcher comes up with an arg `:strict` that affects how floating-point roundoff is dealt with -- then we'd need to hardcode the imatcher arg as `:imatcher_strict` or some other made-up name, adding more confusion for users and developers alike.


     * It introduces unneeded coupling between the gems.  I could easily imagine a future update to this gem that would support other image comparison libraries or let the user provide their own.  For that reason there should be a minimal set of hardcoded args likely common to all such libraries, and all the more esoteric args would go in a library-specific hash.

In summary, I still agree with my original proposal.   What do you think now? :)
